parsl/executors/high_throughput/executor.py

```diff
@@ -86,12 +86,14 @@
 
     address : string
         An address to connect to the main Parsl process which is reachable from the network in which
-        workers will be running. This field expects an IPv4 address (xxx.xxx.xxx.xxx).
+        workers will be running. This field expects an IPv4 address (xxx.xxx.xxx.xxx) (when
+        ``enable_ipv6`` is not set)
         Most login nodes on clusters have several network interfaces available, only some of which
         can be reached from the compute nodes. This field can be used to limit the executor to listen
         only on a specific interface, and limiting connections to the internal network.
         By default, the executor will attempt to enumerate and connect through all possible addresses.
         Setting an address here overrides the default behavior.
+        If ``enable_ipv6`` is set, specify an IPv6 address in brackets, for eg: ``address=[::1]``
```

Comment:
This PR sets the `ZMQ_IPV6` option for every socket, regardless of the value of `enable_ipv6`. Thus, users can always specify an IPv6 address for the `address` argument, which is used for communication between the interchange and worker worker managers. The `enable_ipv6` argument only affects communication between the executor and interchange.

Reply:
This style of using brackets here feels a bit awkward to me: it's a piece of URL from https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2 not the standard RFC4291 format of addresses https://datatracker.ietf.org/doc/html/rfc4291#section-2.2

Looking into the future, does that mean some address fields are going to end up taking RFC4291 standard addressing and some are going to be "paste this URL fragment" style addresses? (for example the parameter that sets UDP address that monitoring reports to?)

Reply:
the presence of these brackets seems to cause some trouble in my environment: this shell command

```
$ cat pytest-parsl/parsltest-current/runinfo/000/submit_scripts/parsl.htex_local.block-0.1732487389.7241297.sh

export JOBNAME=parsl.htex_local.block-0.1732487389.7241297
process_worker_pool.py --debug  -a [2001:db8::2] -p 0 -c 1 -m None --poll 10 --task_port=54325 --result_port=54619 --cert_dir /home/benc/parsl/src/parsl/pytest-parsl/parsltest-current/runinfo/000/htex_local/certificates --logdir=/home/benc/parsl/src/parsl/pytest-parsl/parsltest-current/runinfo/000/htex_local --block_id=0 --hb_period=30  --hb_threshold=120 --drain_period=None --cpu-affinity none  --mpi-launcher=mpiexec --available-accelerators
```

results in a parsed address argument list of:

```
$ grep address pytest-parsl/parsltest-current/runinfo/000/htex_local/block-0/b3cb42e760c3/manager.log 
2024-11-24 22:29:50.151 parsl:887 456 MainThread [INFO]  addresses: :
```

Should I be expecting this to work?

Reply:
Hilariously, because this PR asks the user to specify shell globs and then doesn't make the shell not interpret these as globs, the actual behaviour of what gets passed to process_worker_pool.py is dependent on what particular files you have in your working directory!

For example:

```
benc@parsl-dev-3-12-9310:~/parsl/src/parsl$ echo [2001:db8::2]
[2001:db8::2]
benc@parsl-dev-3-12-9310:~/parsl/src/parsl$ touch 2
benc@parsl-dev-3-12-9310:~/parsl/src/parsl$ echo [2001:db8::2]
2
```

Reply:
The brackets should be not necessary, as pyzmq does not require them.

Re: shell globs, we should add single quotes around the `--addresses` launch command option value.

Reply:
ok, without brackets I can run a test successfully:

```
                address="2001:db8::2",
```

Reply:
if you're going to require valid RFC4291 IP addresses on the submit side, maybe don't need to do any quotation against shell globs - at least, not any more than IPv4 addresses need the same protections.

Reply:
> What's weird to me (and maybe what you were observing?) is requiring the brackets when it's strictly the IP address.

right: 3986 is a standard for *URLs*. RFC 4291 is a standard for how to write *IPv6 addresses*. RFC 3986 incorporates (the predecessor to) RFC 4291 as to what goes inside the brackets. The brackets come along when you start doing URLs, not addresses.

Reply:
(or I guess, URL-like things... eg SSH targets too sometimes)